tracer/src/Datadog.Trace/UserDetails.cs

```diff
@@ -0,0 +1,38 @@
+// <copyright file="UserDetails.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+namespace Datadog.Trace
+{
+    /// <summary>
+    /// A data container class for the users details
+    /// </summary>
+    public class UserDetails
```

Comment:
If users are expected to create `UserDetails` on every request/trace, we should consider making this a `struct` (maybe an immutable `readonly struct`?) to avoid the heap allocation. Since this would not be a tiny struct, we could also make the `UserDetails` parameter in the extension method an `in` parameter, so it's passed by reference.

If users decide to cache the `UserDetails` in the user session, it will end up in the heap, but that's still only one instance per user instead of one per request.

Reply:
+1

Reply:
I agree with the struct part, but I don't think we should use `in` when it's not a `readonly struct`.

From [the docs](https://docs.microsoft.com/en-us/dotnet/csharp/write-safe-efficient-code#avoid-defensive-copies):

> Pass a struct as the argument for an in parameter only if it's declared with the readonly modifier or the method accesses only readonly members of the struct. Otherwise, the compiler must create defensive copies in many situations to ensure that arguments are not mutated. Consider the following example that calculates the distance of a 3D point from the origin:

> Without some language guarantee, the compiler must create a temporary copy of the argument before calling any member not marked with the readonly modifier. The temporary storage is created on the stack, the values of the argument are copied to the temporary storage, and the value is copied to the stack for each member access as the this argument. In many situations, these copies harm performance enough that pass-by-value is faster than pass-by-read-only-reference when the argument type isn't a readonly struct and the method calls members that aren't marked readonly.

Reply:
I think the idea is to use `readonly struct`.

Reply:
I mentioned `readonly struct` but I wasn't sure if we should enforce it. Thank you both for keeping me honest 🎉

Reply:
The trouble with `readonly struct` here is that it makes the API worse IMO. No better than just having parameters passed to the method. Either way would be a breaking change if we want to add an extra parameter.

Struct is still an improvement, we just have to accept the copy I think 🤷‍♂️